### PR TITLE
[MPS] Fix pad corrupting rank > 4 tensors with >= 2^16 inner elements

### LINKED ISSUE #194922: F.pad(mode="constant") with temporal-axis padding silently corrupts ~80% of elements for large 5D tensors on MPS
### 🐛 Describe the bug

`F.pad` with `mode="constant"` along the **temporal (dim 2) axis of a 5D tensor** silently writes wrong values into most of the output tensor on the MPS backend when the padded spatial extent is large. No error is raised; corrupted values look statistically plausible (same distribution), so downstream models silently produce garbage.

This silently corrupted a production video-VAE encoder (ComfyUI MiniMax-H3 VAE) above ~224x480 resolution: latents looked plausible but decoded to flat "tan" videos.

### Environment

- PyTorch 2.13.0, macOS arm64, Apple M4 Max
- Reproduces in fp32 and fp16 (fp16 additionally produces NaNs in the padded output)

### Minimal repro

```python
import torch
import torch.nn.functional as F

torch.manual_seed(3)
x = torch.randn(1, 3, 9, 256, 544)          # 5D, fp32
pad = (0, 0, 0, 0, 2, 0)                    # temporal front-only

a = F.pad(x.to("mps"), pad).cpu()           # MPS
b = F.pad(x, pad)                           # CPU
print(torch.equal(a, b))                    # False (~80% of elements differ)
print(float((a - b).abs().max()))           # ~5-6 (input range is +/-4)
```

### Characterization

Elementwise compare vs CPU (`torch.equal`), `pad=(0,0,0,0,2,0)`, fp32, T=9, C=3:

| spatial size | elems/frame (padded) | bad fraction |
|---|---|---|
| 96x224 | 22K | 0.0% |
| 128x288 | 37K | 0.0% |
| 160x352 | 57K | 0.0% |
| 192x416 | 81K | **74.8%** |
| 224x480 | 109K | **58.4%** |
| 252x540 | 137K | **80.3%** |
| 256x544 | 141K | **79.7%** |
| 352x768 | 272K | **80.5%** |

- Threshold tracks elements per spatial frame, near 2^16 = 65,536
- Damage pattern (256x544): output frame 0 (the newly padded frame) is 100% correct; frame 1 is 4.6% bad; frames 2+ are 94.06% bad. Exactly 131,072 (2^17) elements wrong per frame, 8,192 (2^13) correct — the power-of-2 geometry suggests a tiling/grid bug in the pad kernel
- Reproduces with: back-only `(0,0,0,0,0,2)`, symmetric `(0,0,0,0,1,1)`, any `value`, C=128, T=2..17, fp16 (NaNs)
- Does **not** reproduce with spatial-only constant padding `(2,0,0,0,0,0)` even at 352x768, nor with `mode="reflect"`/`"replicate"` on any axis
- A `conv3d` consuming the corrupted pad output is itself correct; padding on CPU then moving to MPS gives bit-exact conv results
- `.clone()`, `.contiguous()`, or inserting ops between pad and consumer do not change the corrupted values (values are wrong at pad time, not a layout issue)

### Impact

Any MPS user of 3D-conv models with causal/temporal zero-padding at >= ~256px spatial sizes (video VAE encoders, video diffusion) gets silent data corruption. Because the threshold is size-dependent and the pattern is model-code-dependent, it presents to users as "model X breaks at 720p on Mac" rather than as a framework bug.

### Workaround

Replace temporal constant-pad with `torch.cat` of a zero slab (bit-exact vs CPU, verified), or perform the pad on CPU.

Possibly related class of bug: #188359 (silently incorrect conv2d output for kernel size >= 256 on MPS).

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by robertomeroni**
## Issue

Fixes #194922

## Summary

MPSGraph's `padTensor:` and `padGradientWithIncomingGradientTensor:` overwrite the copied input data at rank > 4, mostly with zeros and partly with other wrong values. The shape is correct, so nothing notices: `F.pad(randn(1, 3, 9, 256, 544), (0,0,0,0,2,0))` is garbage in 79% of its elements. The boundary is in elements, not bytes: `max(output last dim, product of output dims after the padded axis) >= 65536`.

`pad_out_template` builds the graph at the operand's own rank, and only the trailing `padding_size / 2` dims are padded, so this folds the leading dims into one batch dim, pads at rank <= 4 (measured correct at any extent), and reshapes back in-graph. It is in the shared template, so it also fixes `reflection_pad3d` and `replication_pad3d`, equally corrupt today, forward and backward. #194944 fixes constant padding only, and applies cleanly on top.

## Checklist

- [x] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [ ] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No.

---
AI assistance was used: Claude Code wrote the patch, the test and this description under my direction; I reviewed every line.

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -11985,6 +11975,25 @@ def test_replication_pad1d_non_contig_out(self):
...
+            grad = torch.randn_like(cpu_out)
+            cpu_out.backward(gradient=grad)
+            mps_out.backward(gradient=grad.to("mps"))
+            self.assertEqual(cpu_x.grad, mps_x.grad)
+
+        for mode in ["constant", "replicate", "reflect"]:
+            helper((1, 2, 3, 256, 256), (0, 0, 0, 0, 2, 1), mode)
+        helper((1, 1, 2, 2, 65536), (2, 1), "constant")
+        helper((1, 1, 2, 2, 65536), (0, 0, 2, 1), "constant")
+
```
Comment (Isalia20) [inline test/test_mps.py]: Please move this to OpInfo tests or somewhere where other devices can be tested

Comment (Isalia20) [reply test/test_mps.py]: Also modes could be made as parametrized test

**Review by malfet — APPROVED** [review]

File: test/test_mps.py (hunk 1 above)
Comment (robertomeroni) [reply test/test_mps.py]: thanks, I will update with parametrization

on moving the test: this bug is inside MPSGraph's `padTensor:`, so only the MPS backend can hit it. Every
other test in `TestPad` is also a CPU-vs-MPS comparison, so this feels like the natural home for it. I'd
prefer to keep it here, but I can add a device-generic version in `test_nn.py` as a follow-up if
you'd rather have the coverage there

I'd avoid OpInfo for it. Reproducing this needs an inner extent of 65536 (~262k elements), and the largest
sample in `sample_inputs_nn_pad` today is `(3, 3, 5, 5)`. That generator feeds five pad OpInfos across all
dtypes and is consumed by ~30 test files, so the shape would be costly everywhere while only MPS can fail
it

Comment (Isalia20) [reply test/test_mps.py]: Yeah, it's inside MPS Graph padTensor but MPSGraph are also metal kernels underneath written by humans so other backends are not immune to errors which could have occurred for such non-usual path cases, thus the need to test other backends as well (If such errors occur, this becomes a great initiative to also fix other backends).

Also, in general, for MPS we are moving away from adding tests in test_mps.py unless we want to test behaviour super specific to MPS (for example we wrote some custom metal kernels for it which are instantiated for certain shapes, you can look at prefill/decode kernels and their tests as an example)

Comment (robertomeroni) [reply test/test_mps.py]: done, moved to `TestNNDeviceType` in `test_nn.py`, next to `test_large_reflect_pad`

I scoped it `onlyAccelerator`, so it runs on MPS and CUDA but skips CPU, since CPU is the reference it
compares against. Happy to drop that if you would rather the CPU case ran as a smoke test

**Review by Isalia20 — APPROVED** [review]
LGTM thanks! let's wait for tests to run

File: test/test_mps.py (hunk 1 above)
Comment (Isalia20) [reply test/test_mps.py]: thanks

**Comment by robertomeroni** [comment]
@Isalia20 thanks! There is one failing test, but looks unrelated

**Comment by MohammadSadeghSalehi** [comment]
Nice fix, and a better one than mine. Folding into the batch dim inside `pad_out_template` keeps the MPSGraph path and picks up `reflection_pad3d` and `replication_pad3d` forward and backward, which I had not checked at all. I have closed #194944.

Your boundary condition is sharper than what I posted on the issue. I only varied the product of the two innermost dims and concluded 65536 there; `max(output last dim, product of output dims after the padded axis)` is the right statement.

One thing worth recording for later, since this works around the fault rather than removing it. I isolated it to `MPSGraph` itself, no PyTorch involved: `padTensor:` with `MPSGraphPaddingModeConstant` writes the constant fill value in place of the input data above the threshold, entirely at twice it. Filed with Apple as FB24527244 with a standalone repro attached. So if the rank <= 4 path ever changes shape, the same fault is still sitting underneath.
